File: frontend/src/types/budget.ts
export type BudgetPeriodType = 'monthly' | 'weekly' | 'yearly' | 'custom';
export type CollaboratorRole = 'viewer' | 'editor';

export interface BudgetCategory {
  id: string;
  budget_id: string;
  category_id: string;
  limit_amount: string;
}

export interface Budget {
  id: string;
  owner_id: string;
  name: string;
  period_type: BudgetPeriodType;
  start_date: string;
  end_date: string | null;
  currency: string;
  budget_categories: BudgetCategory[];
  created_at: string;
  updated_at: string;
}

export interface BudgetCreate {
  name: string;
  period_type: BudgetPeriodType;
  start_date: string;
  end_date?: string | null;
  currency: string;
  budget_categories?: { category_id: string; limit_amount: string }[];
}

export interface BudgetUpdate {
  name?: string;
  period_type?: BudgetPeriodType;
  start_date?: string;
  end_date?: string | null;
  currency?: string;
  budget_categories?: { category_id: string; limit_amount: string }[];
}

export interface Collaborator {
  id: string;
  budget_id: string;
  user_id: string;
  role: CollaboratorRole;
  invited_at: string;
  accepted_at: string | null;
}

export interface CollaboratorInvite {
  email: string;
  role?: CollaboratorRole;
}

export interface CollaboratorUpdate {
  role: CollaboratorRole;
}

export interface BudgetSummaryCategory {
  category_id: string;
  category_name: string;
  limit_amount: string;
  spent_amount: string;
  remaining: string;
}

export interface BudgetSummary {
  budget_id: string;
  budget_name: string;
  period_type: string;
  start_date: string;
  end_date: string | null;
  currency: string;
  categories: BudgetSummaryCategory[];
  total_limit: string;
  total_spent: string;
  total_remaining: string;
}

File: backend/app/schemas/budget.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, EmailStr


class BudgetCategoryCreate(BaseModel):
    category_id: uuid.UUID
    limit_amount: Decimal


class BudgetCategoryResponse(BaseModel):
    id: uuid.UUID
    budget_id: uuid.UUID
    category_id: uuid.UUID
    limit_amount: Decimal

    model_config = {"from_attributes": True}


class BudgetCreate(BaseModel):
    name: str
    period_type: str  # monthly, weekly, yearly, custom
    start_date: date
    end_date: date | None = None
    currency: str
    budget_categories: list[BudgetCategoryCreate] = []


class BudgetUpdate(BaseModel):
    name: str | None = None
    period_type: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    currency: str | None = None


class BudgetResponse(BaseModel):
    id: uuid.UUID
    owner_id: uuid.UUID
    name: str
    period_type: str
    start_date: date
    end_date: date | None
    currency: str
    budget_categories: list[BudgetCategoryResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CollaboratorInvite(BaseModel):
    email: EmailStr
    role: str = "viewer"  # viewer, editor


class CollaboratorUpdate(BaseModel):
    role: str  # viewer, editor


class CollaboratorResponse(BaseModel):
    id: uuid.UUID
    budget_id: uuid.UUID
    user_id: uuid.UUID
    role: str
    invited_at: datetime
    accepted_at: datetime | None

    model_config = {"from_attributes": True}


class BudgetSummaryCategory(BaseModel):
    category_id: uuid.UUID
    category_name: str
    limit_amount: Decimal
    spent_amount: Decimal
    remaining: Decimal


class BudgetSummaryResponse(BaseModel):
    budget_id: uuid.UUID
    budget_name: str
    period_type: str
    start_date: date
    end_date: date | None
    currency: str
    categories: list[BudgetSummaryCategory]
    total_limit: Decimal
    total_spent: Decimal
    total_remaining: Decimal
